cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewRequestQueue.java

```diff
@@ -0,0 +1,134 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.preview;
+
+import com.google.gson.JsonObject;
+import com.google.inject.Inject;
+import io.cdap.cdap.app.preview.PreviewRequest;
+import io.cdap.cdap.app.preview.PreviewRequestQueue;
+import io.cdap.cdap.app.preview.PreviewRequestQueueState;
+import io.cdap.cdap.app.store.preview.PreviewStore;
+import io.cdap.cdap.common.app.RunIds;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.proto.id.ApplicationId;
+import net.jcip.annotations.ThreadSafe;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Thread-safe implementation of {@link PreviewRequestQueue} backed by {@link PreviewStore}.
+ */
+@ThreadSafe
+public class DefaultPreviewRequestQueue implements PreviewRequestQueue {
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultPreviewRequestQueue.class);
+  private final PreviewStore previewStore;
+  private final ConcurrentLinkedDeque<PreviewRequest> requestQueue;
+  private final int capacity;
+  private final long waitTimeOut;
+  private final AtomicInteger queueSize;
+
+  @Inject
+  DefaultPreviewRequestQueue(CConfiguration cConf, PreviewStore previewStore) {
+    this.previewStore = previewStore;
+    this.capacity = cConf.getInt(Constants.Preview.WAITING_QUEUE_CAPACITY, 50);
+    this.waitTimeOut = cConf.getLong(Constants.Preview.WAITING_QUEUE_TIMEOUT_SECONDS, 60);
+    this.requestQueue = new ConcurrentLinkedDeque<>();
+    this.queueSize = new AtomicInteger();
+    List<PreviewRequest> allInWaitingState = previewStore.getAllInWaitingState();
+    for (PreviewRequest request : allInWaitingState) {
+      if (isTimedOut(request)) {
+        continue;
+      }
+      requestQueue.add(request);
+      queueSize.incrementAndGet();
+    }
+  }
+
+  @Override
+  public Optional<PreviewRequest> poll(JsonObject pollerInfo) {
+    while (true) {
+      PreviewRequest previewRequest = requestQueue.poll();
+      if (previewRequest == null) {
+        return Optional.empty();
+      }
+
+      if (isTimedOut(previewRequest)) {
+        LOG.warn("Preview request wth application id {} is timed out. Ignoring it.",
+                 previewRequest.getProgram().getParent());
+        continue;
+      }
+
+      try {
+        previewStore.setPreviewRequestPollerInfo(previewRequest.getProgram().getParent(), pollerInfo);
+      } catch (IllegalArgumentException e) {
+        LOG.warn("Error while setting the poller info for preview request with application id {}. Ignoring the preview",
```

Comment:
Why? This means a misbehaving client can actually causes all preview requests being ignored?

Reply:
This was added if the poller info is attempted to set but the preview is not in WAITING state (for example when killed by user). Throwing `ConflictException` from the  `setPreviewRequestPollerInfo` for this case now to make contract clear. PreviewRequest will be ignored.